caas/kubernetes/provider/resources/statefulset.go

```diff
@@ -48,28 +49,56 @@ func (ss *StatefulSet) ID() ID {
 }
 
 // Apply patches the resource change.
-func (ss *StatefulSet) Apply(ctx context.Context, client kubernetes.Interface) error {
+func (ss *StatefulSet) Apply(ctx context.Context, client kubernetes.Interface) (err error) {
 	api := client.AppsV1().StatefulSets(ss.Namespace)
-	data, err := runtime.Encode(unstructured.UnstructuredJSONScheme, &ss.StatefulSet)
+	var result *appsv1.StatefulSet
+	defer func() {
+		if result != nil {
+			ss.StatefulSet = *result
+		}
+	}()
+
+	result, err = api.Create(ctx, &ss.StatefulSet, metav1.CreateOptions{
+		FieldManager: JujuFieldManager,
+	})
+	if k8serrors.IsAlreadyExists(err) {
+		// Continue to patch.
+	} else if err != nil {
+		return errors.Trace(err)
+	} else {
+		return nil
+	}
+
+	existing, err := api.Get(ctx, ss.Name, metav1.GetOptions{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	existing.SetAnnotations(ss.GetAnnotations())
+	existing.Spec.Replicas = ss.Spec.Replicas
+	existing.Spec.UpdateStrategy = ss.Spec.UpdateStrategy
+	existing.Spec.Template.Spec.Volumes = ss.Spec.Template.Spec.Volumes
+	existing.Spec.Template.SetAnnotations(ss.Spec.Template.GetAnnotations())
+	existing.Spec.Template.Spec.Containers = ss.Spec.Template.Spec.Containers
+	existing.Spec.Template.Spec.ServiceAccountName = ss.Spec.Template.Spec.ServiceAccountName
+	existing.Spec.Template.Spec.AutomountServiceAccountToken = ss.Spec.Template.Spec.AutomountServiceAccountToken
+	// NB: we can't update the Spec.ServiceName as it is immutable.
```

Comment:
This change makes statefulset updates different to all the other k8s resources juju manages. Everything currently uses the same pattern:
1. patch
2. if not found, create

Can we simply alter the existing logic (for everything) to check the error from patching and if it's "ErrConflict", then refresh the resource and try again?